src/macros.rs

```diff
@@ -77,6 +77,52 @@ macro_rules! tinystr16 {
     };
 }
 
+// Internal macro for implementing Serialize/Deserialize
+macro_rules! serde_impl {
+    ($ty:ty, $int:ident) => {
+        #[cfg(feature = "serde")]
+        impl serde::Serialize for $ty {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                if serializer.is_human_readable() {
+                    serializer.serialize_str(self.as_str())
+                } else {
+                    self.as_unsigned().to_le().serialize(serializer)
+                }
+            }
+        }
+
+        #[cfg(feature = "serde")]
+        impl<'de> serde::Deserialize<'de> for $ty {
+            fn deserialize<D>(deserializer: D) -> Result<$ty, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                use std::borrow::Cow;
+                use serde::de::Error as SerdeError;
+                use std::string::ToString;
+
+                if deserializer.is_human_readable() {
+                    let x: Cow<'de, str> = serde::Deserialize::deserialize(deserializer)?;
+                    x.parse().map_err(|e: Error| SerdeError::custom(e.to_string()))
+                } else {
+                    // little-endian
+                    let le = serde::Deserialize::deserialize(deserializer)?;
+                    let bytes = $int::from_le(le).to_ne_bytes();
+                    if let Err(e) = std::str::from_utf8(&bytes) {
+                        return Err(SerdeError::custom(e.to_string()))
+                    }
+                    unsafe {
+                        Ok(Self::new_unchecked(le))
+                    }
```

Comment:
I strongly object to having an `unsafe` hidden inside a deserializer.  Unless we can mark the whole deserializer implementation as `unsafe`, then we need to do more work to verify that the byte sequence is valid for a TinyStr.

In other words, just use `TinyStr::from_bytes` to both validate and construct.